src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -139,24 +169,50 @@ ObjectDataHandler::write_ret do_removals(
   context_t ctx,
   lba_pin_list_t &pins)
 {
-  return trans_intr::do_for_each(
-    pins,
-    [ctx](auto &pin) {
+  return seastar::do_with(
+    pins.front()->get_key(),
+    pins.front()->get_key(),
+    [ctx, &pins](auto &next_laddr, auto &last_laddr) {
+    return trans_intr::do_for_each(
+      pins,
+      [ctx, &next_laddr, &last_laddr](auto &pin) {
       LOG_PREFIX(object_data_handler.cc::do_removals);
+      if (pin->get_key() < next_laddr) {
+	// with lba indirect in place, chances are that
+	// one lba mapping corresponds to multiple pins.
+	// For example:
+	// 	lba_mapping for clone1: laddr_1a ~ 4096, laddr_1b ~ 4096
+	// 	lba_mapping for clone2: laddr_2a ~ 8192(indirection of
+	// 						laddr_2a is laddr_1a)
+	// Although clone2 has only one lba_mapping laddr_2a ~ 8192, transaction
+	// manager's get_pins() will return two pins: laddr_2a~4096,laddr_2a+4096~4096
```

Comment:
> If that's true, how is this possible? Seems to me that the symmetric cloning [#51141 (comment)](https://github.com/ceph/ceph/pull/51141#discussion_r1190683853) can only make two clones pointing to a shared indirect laddr+len, but cannot make one clone pointing to another clone, right?

Sorry, this comment is misleading, by `clone 1`, I actually meant the the data that are shared by clones. I'll corret it.

Reply:
I'm moving the disussion https://github.com/ceph/ceph/pull/51141#issuecomment-1571762097 here so it can be chained, and I think this is related about how we should handle extent split/overwrite/remove with indirection.

> 2. the head's version of E needs to be splitted, E also needs to be splitted(this may not be necessary once the extent partial read is merged)

There is an assumption that all the clones are symmetric from SeaStore point-of-view, and they should support modifications independently. When a clone splits its extent, the indirect/shared extent also will split, even if it is shared by other clones. I'm not sure how the ref-counter is maintained with this design.

For example, there are `n`(2+) clones (`laddr_c1 ~ 16K`, `laddr_c2 ~ 16K`, ...) sharing the same indirect data `(laddr_i ~ 16K)[n]`, [n] means the ref-counter. In the following steps:
* c1 overwrites `laddr_c1 + 4K ~ 4K`
* c2 overwrites `laddr_c2 + 8K ~ 4K`

How does the indirect data `(laddr_i ~ 16K)[n]` splitted with the ref-counter, so that removing any of the n clones will be correct?

Reply:
> E also needs to be splitted(this may not be necessary once the extent partial read is merged)

Seems to me this is not dependent to whether extent partial read is merged, feature-wise, as it just need to fallback to read the whole extent in the current version. If not-splitting is the correct direction, why need to take a detour with splitting? We need to spend extra efforts to make sure splitting is correct and then need to remove&replace it correctly.

Reply:
> For example, there are `n`(2+) clones (`laddr_c1 ~ 16K`, `laddr_c2 ~ 16K`, ...) sharing the same indirect data `(laddr_i ~ 16K)[n]`, [n] means the ref-counter. In the following steps:
> 
> * c1 overwrites `laddr_c1 + 4K ~ 4K`
> * c2 overwrites `laddr_c2 + 8K ~ 4K`
> 
> How does the indirect data `(laddr_i ~ 16K)[n]` splitted with the ref-counter, so that removing any of the n clones will be correct?

This would result in the following mappings:
(laddr_i, 4K)[n], (laddr_i + 4K, 4K)[n-1], (laddr_i + 8K, 4K)[n-1], (laddr_i + 12K, 4K)[n]

Reply:
> Seems to me this is not dependent to whether extent partial read is merged, feature-wise, as it just need to fallback to read the whole extent in the current version.

Let me think about it.

Reply:
> > Seems to me this is not dependent to whether extent partial read is merged, feature-wise, as it just need to fallback to read the whole extent in the current version.
> 
> Let me think about it, I did consider this approach, but I don't quite recall the conclusion:-)

@cyx1231st I recalled it.

Consider the following scenario:

1. we have an extent `[laddr, 16K]` that are shared by `clone1([laddr_c1, 16K)` and `head[laddr_h, 16K]`;
2. `Transaction A` overwrite [laddr_h, 8K];
3. `Transaction B` creates `clone2` and overwrite [laddr_h + 8K, 8K]

`Transaction B` should decrease the ref count of `[laddr, 16K]`. However, there's no way for `Transaction B` to know that there had been a previous transaction that has overwritten `[laddr_h, 8K]`, so `Transaction B` can hardly do the ref decrement, unless it examine all the previous clone onodes.

Reply:
Yeah, I was also thinking about it. Modifying the ref-counter will mutate the extent, which should be correct to generate transaction conflicts.